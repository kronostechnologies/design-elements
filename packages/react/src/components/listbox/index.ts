export { Listbox } from './listbox';
export type { ListboxOption, ListboxProps } from './listbox';
