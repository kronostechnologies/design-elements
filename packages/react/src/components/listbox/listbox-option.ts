type Value = string | string[];

export function findOptionsByValue<T extends { value: string }>(
    options: T[],
    searchValue?: Value | undefined,
): T[] {
    return options.filter(
        (option) => (Array.isArray(searchValue)
            ? searchValue.includes(option.value)
            : option.value === searchValue),
    );
}
