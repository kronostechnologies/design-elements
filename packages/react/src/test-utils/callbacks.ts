export function doNothing(): void {
    // do nothing
}
