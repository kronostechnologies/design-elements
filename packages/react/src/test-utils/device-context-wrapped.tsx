import React, { ReactElement } from 'react';
import { DeviceContextProvider, DeviceType } from '../components/device-context-provider/device-context-provider';

export function DeviceContextWrapped(children: ReactElement, device?: DeviceType): ReactElement {
    return (
        <DeviceContextProvider staticDevice={device}>
            {children}
        </DeviceContextProvider>
    );
}
