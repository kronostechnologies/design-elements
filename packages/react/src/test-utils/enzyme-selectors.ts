import { ReactWrapper, ShallowWrapper } from 'enzyme';

enum ModifierEnum {
    StartsWith = '^',
    EndsWith = '$',
    ContainsOptionalHyphen = '|',
    Contains = '*',
    ContainsWord = '~',
    None = '',
}
type SelectorModifier = `${ModifierEnum}`;

function filterHtmlNodesOnly<W extends ReactWrapper | ShallowWrapper>(
    nodes: W,
): W {
    return nodes
        .hostNodes()
        .filterWhere((x) => x.name() !== null && !x.name().startsWith('styled.')) as W;
}

export function findByTestId<W extends ReactWrapper | ShallowWrapper>(
    wrapper: W,
    testId: string,
    options?: {
        modifier?: SelectorModifier,
        nodeName?: string,
        // Don't use with shallow wrapper as it contains styled node over the html nodes.
        htmlNodesOnly?: boolean,
    },
): W {
    const modifier = options?.modifier ?? '';

    let nodes = wrapper.find(`[data-testid${modifier}="${testId}"]`) as W;

    if (options?.nodeName) {
        nodes = nodes.filterWhere((x) => x.name() === options.nodeName) as W;
    }
    if (options?.htmlNodesOnly) {
        nodes = filterHtmlNodesOnly(nodes);
    }

    return nodes;
}

export function getByTestId<W extends ReactWrapper | ShallowWrapper>(
    wrapper: W,
    testId: string,
    options?: {
        modifier?: SelectorModifier,
        nodeName?: string,
        // Don't use with shallow wrapper as it contains styled node instead of the html nodes.
        htmlNodesOnly?: boolean,
    },
): W {
    return findByTestId(wrapper, testId, options).at(0) as W;
}
