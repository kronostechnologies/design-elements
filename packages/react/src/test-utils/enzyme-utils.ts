import { ReactWrapper } from 'enzyme';
import { act } from 'react-dom/test-utils';

export async function waitForComponentToPaint<P = {}>(
    wrapper: ReactWrapper<P>,
    amount = 0,
): Promise<void> {
    await act(async () => {
        await new Promise((resolve) => setTimeout(resolve, amount));
        wrapper.update();
    });
}

export function expectFocusToBeOn(element: ReactWrapper): void {
    expect(document.activeElement).toBe(element.getDOMNode());
}
