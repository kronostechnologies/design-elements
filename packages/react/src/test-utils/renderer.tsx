import { ThemeWrapped } from '@design-elements/test-utils/theme-wrapped';
import { mount, ReactWrapper, render } from 'enzyme';
import React, { Component, ReactElement } from 'react';
import { MemoryRouter } from 'react-router-dom';
import { DeviceContextProvider } from '../components/device-context-provider/device-context-provider';

export function mountWithProviders<C extends Component, P = C['props'], S = C['state']>(
    component: ReactElement<P>,
): ReactWrapper<P, S, C> {
    return mount(component, {
        wrappingComponent: AllProviders,
    });
}

export function renderWithProviders(
    component: ReactElement,
): Cheerio {
    return render(<AllProviders>{component}</AllProviders>);
}

function AllProviders({ children }: { children: ReactElement }): ReactElement {
    return (
        <MemoryRouter>
            <DeviceContextProvider>
                {ThemeWrapped(children)}
            </DeviceContextProvider>
        </MemoryRouter>
    );
}
