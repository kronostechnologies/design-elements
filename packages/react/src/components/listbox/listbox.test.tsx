import { getByTestId } from '@design-elements/test-utils/enzyme-selectors';
import { mountWithTheme, renderWithTheme, shallowWithTheme } from '@design-elements/test-utils/renderer';
import React from 'react';
import { Listbox } from './listbox';

jest.mock('@design-elements/utils/uuid');

const options = [
    {
        label: 'Option A',
        value: 'optionA',
    },
    {
        label: 'Option B',
        value: 'optionB',
    },
    {
        label: 'Option C',
        value: 'optionC',
    },
    {
        label: 'Option D',
        value: 'optionD',
    },
];

describe('Listbox', () => {
    test('Correctly set the default option', () => {
        const wrapper = shallowWithTheme(<Listbox options={options} defaultValue="optionB" />);

        expect(getByTestId(wrapper, 'listitem-optionB').prop('selected')).toEqual(true);
    });

    test('Correctly set the default option with array', () => {
        const wrapper = shallowWithTheme(<Listbox options={options} defaultValue={['optionB']} />);

        expect(getByTestId(wrapper, 'listitem-optionB').prop('selected')).toEqual(true);
    });

    test('Correctly set the default option to the first array element given multiselect is set to false', () => {
        const wrapper = shallowWithTheme(
            <Listbox options={options} defaultValue={['optionA', 'optionB']} multiselect={false} />,
        );

        expect(getByTestId(wrapper, 'listitem-optionA').prop('selected')).toEqual(true);
        expect(getByTestId(wrapper, 'listitem-optionB').prop('selected')).toEqual(false);
    });

    test('Correctly set the default options given multiselect is set to true', () => {
        const wrapper = shallowWithTheme(
            <Listbox options={options} defaultValue={['optionA', 'optionB']} multiselect />,
        );

        expect(getByTestId(wrapper, 'listitem-optionA').prop('selected')).toEqual(true);
        expect(getByTestId(wrapper, 'listitem-optionB').prop('selected')).toEqual(true);
    });

    test('Calls onChange callback when an option is selected', () => {
        const callback = jest.fn();
        const wrapper = shallowWithTheme(<Listbox options={options} defaultValue="optionB" onChange={callback} />);

        getByTestId(wrapper, 'listitem-optionC').simulate('click');

        expect(callback).toHaveBeenCalledTimes(1);
    });

    test('Options should not be selected when no defaultValue is provided', () => {
        const wrapper = shallowWithTheme(<Listbox options={options} />);

        expect(getByTestId(wrapper, 'listitem-optionA').prop('selected')).toBe(false);
        expect(getByTestId(wrapper, 'listitem-optionB').prop('selected')).toBe(false);
        expect(getByTestId(wrapper, 'listitem-optionC').prop('selected')).toBe(false);
        expect(getByTestId(wrapper, 'listitem-optionD').prop('selected')).toBe(false);
    });

    test('Should have the check indicator if selected', () => {
        const wrapper = shallowWithTheme(<Listbox options={options} checkIndicator />);

        expect(getByTestId(wrapper, 'listitem-optionC').find('[data-testid="check-icon"]')).toHaveLength(0);

        getByTestId(wrapper, 'listitem-optionC').simulate('click');

        expect(getByTestId(wrapper, 'listitem-optionC').find('[data-testid="check-icon"]')).toHaveLength(1);
    });

    test('Should allow multiple selections given multiselect is set to true', () => {
        const wrapper = shallowWithTheme(<Listbox options={options} defaultValue={['optionA']} multiselect />);

        getByTestId(wrapper, 'listitem-optionC').simulate('click');

        expect(getByTestId(wrapper, 'listitem-optionA').prop('selected')).toBe(true);
        expect(getByTestId(wrapper, 'listitem-optionC').prop('selected')).toBe(true);
    });

    test('Should unselect option given multiselect is set to true and option is selected', () => {
        const wrapper = shallowWithTheme(<Listbox options={options} defaultValue={['optionA']} multiselect />);

        getByTestId(wrapper, 'listitem-optionA').simulate('click');

        expect(getByTestId(wrapper, 'listitem-optionA').prop('selected')).toBe(false);
    });

    test('Should update value given value prop changed', () => {
        const wrapper = mountWithTheme(<Listbox options={options} value="optionA" />);

        wrapper.setProps({ value: 'optionB' }).update();

        expect(getByTestId(wrapper, 'listitem-optionA').props().selected).toBe(false);
        expect(getByTestId(wrapper, 'listitem-optionB').props().selected).toBe(true);
    });

    test('Should not update value given value prop is undefined', () => {
        const wrapper = mountWithTheme(<Listbox options={options} value="optionA" />);

        wrapper.setProps({ value: undefined }).update();

        expect(getByTestId(wrapper, 'listitem-optionA').props().selected).toBe(true);
    });

    test('Matches the snapshot', () => {
        const tree = renderWithTheme(
            <Listbox options={options} defaultValue="optionB" checkIndicator numberOfItemsVisible={3} />,
        );

        expect(tree).toMatchSnapshot();
    });

    test('Matches the snapshot (multiselect + checkIndicator)', () => {
        const tree = renderWithTheme(
            <Listbox options={options} defaultValue={['optionA', 'optionC']} checkIndicator multiselect />,
        );

        expect(tree).toMatchSnapshot();
    });

    test('Matches the snapshot (dropdown)', () => {
        const tree = renderWithTheme(
            <Listbox options={options} defaultValue="optionB" checkIndicator numberOfItemsVisible={3} dropdown />,
        );

        expect(tree).toMatchSnapshot();
    });
});
