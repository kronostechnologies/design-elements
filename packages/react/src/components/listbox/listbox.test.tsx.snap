// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Listbox Matches the snapshot (multiselect) 1`] = `
.c0 {
  background-color: #FFFFFF;
  border: 1px solid #878F9A;
  border-radius: var(--border-radius);
  box-shadow: 0 0 0 1px #DBDEE1,0 10px 20px 0 rgb(0 0 0 / 0.2);
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  max-height: 160px;
  overflow-y: auto;
  padding: var(--spacing-half) 0;
  position: relative;
  z-index: 1000;
}

.c0 {
  outline: 2px solid transparent;
  outline-offset: -2px;
}

.c0:focus {
  box-shadow: 0 0 0 2px #006296;
  outline: 2px solid #84C6EA;
  outline-offset: -2px;
}

.c1 {
  height: 100%;
  list-style-type: none;
  margin: 0;
  padding: 0;
  width: 100%;
}

.c6 {
  color: #FFFFFF;
  height: 100%;
  width: 100%;
}

.c4 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background-color: #FFFFFF;
  border: 1px solid #60666E;
  border-radius: var(--border-radius);
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: var(--size-1x);
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  margin-right: var(--spacing-1x);
  width: var(--size-1x);
}

.c4:hover {
  background-color: #DBDEE1;
  border: 1px solid #1B1C1E;
}

.c10 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background-color: #FFFFFF;
  border: 1px solid #60666E;
  border-radius: var(--border-radius);
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: var(--size-1x);
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  margin-right: var(--spacing-1x);
  width: var(--size-1x);
}

.c10:hover {
  background-color: #DBDEE1;
  border: 1px solid #1B1C1E;
}

.c10 > .c5 {
  display: none;
}

.c12 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background-color: #F1F2F2;
  border: 1px solid #DBDEE1;
  border-radius: var(--border-radius);
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: var(--size-1x);
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  margin-right: var(--spacing-1x);
  width: var(--size-1x);
}

.c12:hover {
  background-color: #F1F2F2;
  border: 1px solid #DBDEE1;
}

.c12 > .c5 {
  display: none;
}

.c2 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  color: #000000;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  font-size: 0.875rem;
  font-weight: var(--font-semi-bold);
  line-height: var(--size-1halfx);
  min-height: var(--size-1halfx);
  padding: var(--spacing-half) var(--spacing-2x);
  position: relative;
  padding-right: var(--spacing-1x);
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.c2:hover {
  background-color: #DBDEE1;
}

.c2 .c3 {
  background-color: #006296;
  border: 1px solid #006296;
}

.c9 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  color: #000000;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  font-size: 0.875rem;
  font-weight: var(--font-normal);
  line-height: var(--size-1halfx);
  min-height: var(--size-1halfx);
  padding: var(--spacing-half) var(--spacing-2x);
  position: relative;
  padding-right: var(--spacing-1x);
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.c9:hover {
  background-color: #DBDEE1;
}

.c11 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  color: #B7BBC2;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  font-size: 0.875rem;
  font-weight: var(--font-normal);
  line-height: var(--size-1halfx);
  min-height: var(--size-1halfx);
  padding: var(--spacing-half) var(--spacing-2x);
  position: relative;
  padding-right: var(--spacing-1x);
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.c11:hover {
  background-color: #FFFFFF;
}

.c7 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c8 {
  color: #60666E;
  display: block;
  font-size: 0.75rem;
}

<div
  aria-multiselectable="true"
  class="c0"
  data-testid="listbox-container"
  id="uuid1"
  role="listbox"
  tabindex="0"
>
  <ul
    class="c1"
    data-testid="listbox-list"
    role="presentation"
  >
    <li
      aria-selected="true"
      class="c2"
      data-testid="listitem-optionA"
      id="uuid1_optionA"
      role="option"
    >
      <span
        aria-hidden="true"
        checked=""
        class="c3 c4"
      >
        <svg
          class="c5 c6"
          color="currentColor"
          focusable="false"
          height="24"
          width="24"
        />
      </span>
      <span
        class="c7"
      >
        Option A
        <span
          class="c8"
        >
          The first one
        </span>
      </span>
    </li>
    <li
      class="c9"
      data-testid="listitem-optionB"
      id="uuid1_optionB"
      role="option"
    >
      <span
        aria-hidden="true"
        class="c3 c10"
      >
        <svg
          class="c5 c6"
          color="currentColor"
          focusable="false"
          height="24"
          width="24"
        />
      </span>
      <span
        class="c7"
      >
        Option B
      </span>
    </li>
    <li
      aria-selected="true"
      class="c2"
      data-testid="listitem-optionC"
      id="uuid1_optionC"
      role="option"
    >
      <span
        aria-hidden="true"
        checked=""
        class="c3 c4"
      >
        <svg
          class="c5 c6"
          color="currentColor"
          focusable="false"
          height="24"
          width="24"
        />
      </span>
      <span
        class="c7"
      >
        Option C
      </span>
    </li>
    <li
      aria-disabled="true"
      class="c11"
      data-testid="listitem-optionD"
      id="uuid1_optionD"
      role="option"
    >
      <span
        aria-hidden="true"
        class="c3 c12"
        disabled=""
      >
        <svg
          class="c5 c6"
          color="currentColor"
          focusable="false"
          height="24"
          width="24"
        />
      </span>
      <span
        class="c7"
      >
        Option D
      </span>
    </li>
    <li
      class="c9"
      data-testid="listitem-optionE"
      id="uuid1_optionE"
      role="option"
    >
      <span
        aria-hidden="true"
        class="c3 c10"
      >
        <svg
          class="c5 c6"
          color="currentColor"
          focusable="false"
          height="24"
          width="24"
        />
      </span>
      <span
        class="c7"
      >
        Option E
      </span>
    </li>
  </ul>
</div>
`;

exports[`Listbox Matches the snapshot 1`] = `
.c0 {
  background-color: #FFFFFF;
  border: 1px solid #878F9A;
  border-radius: var(--border-radius);
  box-shadow: 0 0 0 1px #DBDEE1,0 10px 20px 0 rgb(0 0 0 / 0.2);
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  max-height: 160px;
  overflow-y: auto;
  padding: var(--spacing-half) 0;
  position: relative;
  z-index: 1000;
}

.c0 {
  outline: 2px solid transparent;
  outline-offset: -2px;
}

.c0:focus {
  box-shadow: 0 0 0 2px #006296;
  outline: 2px solid #84C6EA;
  outline-offset: -2px;
}

.c1 {
  height: 100%;
  list-style-type: none;
  margin: 0;
  padding: 0;
  width: 100%;
}

.c2 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  color: #000000;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  font-size: 0.875rem;
  font-weight: var(--font-normal);
  line-height: var(--size-1halfx);
  min-height: var(--size-1halfx);
  padding: var(--spacing-half) var(--spacing-2x);
  position: relative;
  padding-right: var(--spacing-1x);
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.c2:hover {
  background-color: #DBDEE1;
}

.c5 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  color: #000000;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  font-size: 0.875rem;
  font-weight: var(--font-semi-bold);
  line-height: var(--size-1halfx);
  min-height: var(--size-1halfx);
  padding: var(--spacing-half) var(--spacing-2x);
  position: relative;
  padding-right: var(--spacing-1x);
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.c5:hover {
  background-color: #DBDEE1;
}

.c5::before {
  content: '';
  display: block;
  height: 100%;
  left: 0;
  position: absolute;
  width: 4px;
}

.c6 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  color: #B7BBC2;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  font-size: 0.875rem;
  font-weight: var(--font-normal);
  line-height: var(--size-1halfx);
  min-height: var(--size-1halfx);
  padding: var(--spacing-half) var(--spacing-2x);
  position: relative;
  padding-right: var(--spacing-1x);
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.c6:hover {
  background-color: #FFFFFF;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c4 {
  color: #60666E;
  display: block;
  font-size: 0.75rem;
}

<div
  class="c0"
  data-testid="listbox-container"
  id="uuid1"
  role="listbox"
  tabindex="0"
>
  <ul
    class="c1"
    data-testid="listbox-list"
    role="presentation"
  >
    <li
      class="c2"
      data-testid="listitem-optionA"
      id="uuid1_optionA"
      role="option"
    >
      <span
        class="c3"
      >
        Option A
        <span
          class="c4"
        >
          The first one
        </span>
      </span>
    </li>
    <li
      aria-selected="true"
      class="c5"
      data-testid="listitem-optionB"
      id="uuid1_optionB"
      role="option"
    >
      <span
        class="c3"
      >
        Option B
      </span>
    </li>
    <li
      class="c2"
      data-testid="listitem-optionC"
      id="uuid1_optionC"
      role="option"
    >
      <span
        class="c3"
      >
        Option C
      </span>
    </li>
    <li
      aria-disabled="true"
      class="c6"
      data-testid="listitem-optionD"
      id="uuid1_optionD"
      role="option"
    >
      <span
        class="c3"
      >
        Option D
      </span>
    </li>
    <li
      class="c2"
      data-testid="listitem-optionE"
      id="uuid1_optionE"
      role="option"
    >
      <span
        class="c3"
      >
        Option E
      </span>
    </li>
  </ul>
</div>
`;
